### PR TITLE
feat(argo-cd): Add argocd-cmd-params-cm

### PR DESCRIPTION
**Pull request by pdrastil**
This PR should start discussion regarding https://github.com/argoproj/argo-helm/issues/1210

Changes:
- Added argocd-cmd-params-cm configmap
- Exposed all known config options via environment variables
- Rewire internal parameters for Redis / Dex
- Added deprecation notice for arguments that were superseded by new config

Resolves:
- https://github.com/argoproj/argo-helm/issues/1210

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo).
* [x] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

Changes are automatically published when merged to `master`. They are not published on branches.

### CODE REVIEW

File: charts/argo-cd/README.md.gotmpl
Comment (pdrastil) [inline charts/argo-cd/README.md.gotmpl:44]: Shouldn't this be also changed to `replicas` as everywhere else?

Comment (mkilchhofer) [reply charts/argo-cd/README.md.gotmpl:44]: I think both new components of 2.3 (notification & appset controller) does not yet support HA?

File: charts/argo-cd/templates/argocd-configs/argocd-cmd-params-cm.yaml
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-configs/argocd-cmd-params-cm.yaml:24]: I think we should prepare a breaking change (major version) and refactor it. It looks kinda strange to do a random variable mapping here. Eg. `controller.status.processors = controller.args.statusProcessors`


I'd prefer a structure like:
```yaml
configs:
  cm: # or config: / configmap: / ...?
    # <keys that goes into argocd-cm>

  params:
    # <keys that goes into argocd-cmd-params-cm>
```
Maybe we can use a template function for that and merge the resulting yaml with our templated things like 

- `dex.server: "http://{{ template "argo-cd.dex.fullname" . }}:{{ .Values.dex.servicePortHttp }}"`
- `  redis.server: "{{ template "argo-cd.redis.fullname" . }}:{{ .Values.redis.servicePort }}"`
- etc

**Review by mkilchhofer — COMMENTED** [review]
I really like the intent of implementing #1210 but I'd think about refactoring the chart completely as there are already deprecations in your proposal.

File: charts/argo-cd/templates/argocd-configs/argocd-cmd-params-cm.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-configs/argocd-cmd-params-cm.yaml:24]: I'm open about such change. If I understand correctly you would like to consolidate everything under `configs` section including stuff that is currently under `server`? And after that merge user config with internally set keys such as dex, redis, etc.?

File: charts/argo-cd/values.yaml
Comment (pdrastil) [inline charts/argo-cd/values.yaml:1708]: What do you want to do with `cache` and `repoServer` config? I've seen that this is cross-cutting concern where same value repeats for multiple components and I'm not sure if we want to somehow simplify this or we want to keep it dry and somehow merge / template 1 section into params as well.

**Comment by pdrastil** [comment]
@mkilchhofer Please take a look on another iteration and tell me WHYT. When you  wanted to go with breaking change I've tried to consolidate parameters such as `tlsCerts` and `tlsCertsAnnotations` under one config section. Things I haven't visited yet:

- credentials / clusters / templates - Do we want to have special config section for these? What to do with legacy configuration option that is deprecated? remove this one?
- notifications - it have configuration in `argocd-notifications-cm` in upstream repo - maybe it should be also refactored and added to configs
- `configs.cm.url` - could we use this for notification context and maybe ingress host as a default?

**Comment by mkilchhofer** [comment]
Asked the community in the slack channel:
https://cloud-native.slack.com/archives/C02097VUKUM/p1654880501143779

**Comment by pdrastil** [comment]
> Asked the community in the slack channel for their thoughts about this refactoring: https://cloud-native.slack.com/archives/C02097VUKUM/p1654880501143779
> 
> > grafik

Thanks @mkilchhofer for the info - refactoring is not fully done because I'm waiting for the feedback. I plan to finish remaining sections as well if we are fine with the way this PR is heading.

**Review by ArgTang — CHANGES_REQUESTED** [review]
missing migration guide in readme. 
* it would be nice if one is able to override a components loglevel for debugging. Here you can only set the loglevel globally

File: charts/argo-cd/templates/argocd-repo-server/deployment.yaml
Comment (pdrastil) [inline charts/argo-cd/templates/argocd-repo-server/deployment.yaml:71]: @ArgTang Log level / format is now controlled via config map. Just provide proper key in params section for given component.

File: charts/argo-cd/templates/argocd-applicationset/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-applicationset/deployment.yaml:41]: We can align this.

File: charts/argo-cd/values.yaml
Comment (pdrastil) [reply charts/argo-cd/values.yaml:25]: Waiting for 2.4.x release - I've seen that upstream added extra flags so these components should be configurable via env, but now it's possible only via command line args.

File: charts/argo-cd/values.yaml
Comment (pdrastil) [reply charts/argo-cd/values.yaml:28]: @mkilchhofer WDYT? Keep current info level or change to warn?

File: charts/argo-cd/values.yaml
Comment (ArgTang) [reply charts/argo-cd/values.yaml:25]: imo it would be weird if the global logsetting dont work on some components. I would prefer using them as cmd flags if you want to release this before upstream additions.

File: charts/argo-cd/templates/argocd-repo-server/deployment.yaml
Comment (ArgTang) [reply charts/argo-cd/templates/argocd-repo-server/deployment.yaml:71]: ah you are right. thanks for the correction :)

File: charts/argo-cd/templates/argocd-applicationset/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-applicationset/deployment.yaml]: @ArgTang Added to refactoring.

File: charts/argo-cd/templates/argocd-server/clusterrole.yaml
Comment (pdrastil) [inline charts/argo-cd/templates/argocd-server/clusterrole.yaml:30]: @mkilchhofer Why was this added into Cluster admin role for Argo CD? Do we expect that admin shouldn't be able to check logs?

File: charts/argo-cd/values.yaml
Comment (mkilchhofer) [reply charts/argo-cd/values.yaml:28]: Hmmmm I'd keep info and let the user choice :)

File: charts/argo-cd/templates/argocd-server/clusterrole.yaml
Comment (mkilchhofer) [reply charts/argo-cd/templates/argocd-server/clusterrole.yaml:30]: It will be turned on by in future releases. Argoproj core team decided to disable this by default to not break users setups.
And also in my current position at Swiss Post, our security engineering decided to not enabled that capability on prod clusters.

File: charts/argo-cd/README.md.gotmpl
Comment (mkilchhofer) [inline charts/argo-cd/README.md.gotmpl:31]: 2 replica for the application controller is a bad default. The controller cannot shard its work if Argo CD manages only one cluster.

The shard number of a managed cluster is calculated via hostname:
https://github.com/argoproj/argo-cd/blob/v2.4.0/cmd/util/cluster.go#L133

File: charts/argo-cd/README.md.gotmpl
Comment (mkilchhofer) [inline charts/argo-cd/README.md.gotmpl:54]: use 1 as default. See other comment

File: charts/argo-cd/README.md.gotmpl
Comment (mkilchhofer) [inline charts/argo-cd/README.md.gotmpl:118]: Maybe it's time to analyze the other charts and consolidate the naming. I merged one PR for rollouts with extraEnv couple of hours ago :(

File: charts/argo-cd/values.yaml
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:1659]: Yes, why not removing unrelevant settings. The less examples, the less work needed to have a look into when there's a new release coming in.

File: charts/argo-cd/values.yaml
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:1814]:
```suggestion
    extra: {}
```

What about unify how empty dicts are defined across this file?

File: charts/argo-cd/README.md.gotmpl
Comment (pdrastil) [reply charts/argo-cd/README.md.gotmpl:118]: Agree one same naming convention. All original components (server, reposerver and controller) were using env and envFrom so I tried to stick with that. If you want to change these to extraEnv it is also fine. Imo I believe env will be used less often because of params-cm.

File: charts/argo-cd/templates/argocd-server/clusterrole.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-server/clusterrole.yaml:30]: I see - it makes sense from this perspective.

File: charts/argo-cd/README.md.gotmpl
Comment (pdrastil) [reply charts/argo-cd/README.md.gotmpl:31]: I guess number of replicas comes from this example: https://argo-cd.readthedocs.io/en/stable/operator-manual/high_availability/#argocd-application-controller. Do you want to use 3 for HA setups?

File: charts/argo-cd/README.md.gotmpl
Comment (pdrastil) [reply charts/argo-cd/README.md.gotmpl:54]: 1 is not actually great for HA example

Comment (mkilchhofer) [reply charts/argo-cd/README.md.gotmpl:54]: The controller simply does not support HA.
Ref: https://github.com/argoproj/argo-cd/pull/8340/files?diff=unified&w=0&short_path=5ab05fa#diff-5ab05fafb9f84e2256b1b32e91f4de69e89e5eba19eae3818c7aba539fb22fb0

> Currently, the application-controller does not provide high availability to support workload distribution. This causes conflicts and leads to orphaned/unmanaged applications in scenarios where the application controller handling these resources malfunctions/crashes.

File: charts/argo-cd/values.yaml
Comment (pdrastil) [reply charts/argo-cd/values.yaml]: Done

File: charts/argo-cd/README.md.gotmpl
Comment (pdrastil) [reply charts/argo-cd/README.md.gotmpl]: Good to know - I always thought it can handle failure. Anyway updated to 1

File: charts/argo-cd/README.md.gotmpl
Comment (mkilchhofer) [reply charts/argo-cd/README.md.gotmpl:118]: It's almost 50:50 how the current usage is:
```console
mkilchhofer@Kicm-MBP:~/git/github/argo-helm$ git grep Env: | grep values.yaml
charts/argo-cd/values.yaml:  extraEnv: []
charts/argo-cd/values.yaml:  extraEnv: []
charts/argo-rollouts/values.yaml:  extraEnv: []
charts/argo-rollouts/values.yaml:  extraEnv: []
charts/argo-workflows/values.yaml:  extraEnv: []
charts/argo-workflows/values.yaml:  extraEnv: []
charts/argocd-image-updater/values.yaml:extraEnv: []

mkilchhofer@Kicm-MBP:~/git/github/argo-helm$ git grep env: | grep values.yaml
charts/argo-cd/ci/ha-static-values.yaml:  env:
charts/argo-cd/values.yaml:  env:
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  env: []
charts/argo-cd/values.yaml:  #   env:
charts/argo-cd/values.yaml:  env: []
charts/argo-events/values.yaml:  env: []
charts/argo-events/values.yaml:  env: []
charts/argo-workflows/values.yaml:  env: {}
charts/argo-workflows/values.yaml:  env: {}
charts/argocd-image-updater/values.yaml:    #   credentials: env:REGISTRY_SECRET
```

But its really inconsistent 😿 even in one chart it is different 🤣

**Comment by zestrells** [comment]
Hi everyone, is someone taking a look at adding this config `argocd-cmd-params-cm.yaml` to repo-server? We are getting ready to leverage argocd in production and want to ensure we can change the `reposerver.parallelism.limit` value.

File: charts/argo-cd/README.md.gotmpl
Comment (mkilchhofer) [reply charts/argo-cd/README.md.gotmpl:118]: Lets use `env:` and `envFrom:`

**Comment by pdrastil** [comment]
> Hi everyone, is someone taking a look at adding this config `argocd-cmd-params-cm.yaml` to repo-server? We are getting ready to leverage argocd in production and want to ensure we can change the `reposerver.parallelism.limit` value.

@zestrells I'm currently pulling backward comapatible changes from this PR and trying to push it in smaller pieces.

**Comment by jmeridth** [comment]
@pdrastil #1320 has been merged.  `controller.args.appHardResyncPeriod` can be added as deprecated to this PR.  Thank you.

File: charts/argo-cd/templates/NOTES.txt
Comment (jmeridth) [inline charts/argo-cd/templates/NOTES.txt:10]: ❤️

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml
Comment (jmeridth) [inline charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:53]: ❤️

**Review by jmeridth — COMMENTED** [review]
Thank you for adding this

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-server/deployment.yaml:383]:
```suggestion
        {{- toYaml . | nindent 6 }}
```

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-server/deployment.yaml:37]: Not a blocker but the container-level security context is defined like this
```suggestion
      securityContext:
        {{- toYaml . | nindent 8 }}
```

File: charts/argo-cd/templates/argocd-repo-server/deployment.yaml
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-repo-server/deployment.yaml:64]: Please reduce the indent of the env block by 2. We should use the same style inside one yaml file. All lists are defined like this:
```yaml
key:
- listitem1
- listitem2
```
and not like this:
```yaml
key:
  - listitem1
  - listitem2
```

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-server/deployment.yaml:60]: Reduce indent block by 2

File: charts/argo-cd/values.yaml
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:125]:
```suggestion
  # -- DEPRECATED Application controller commandline flags
```

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-server/deployment.yaml:60]: Done

File: charts/argo-cd/templates/argocd-repo-server/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-repo-server/deployment.yaml]: Done

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-server/deployment.yaml]: Done

File: charts/argo-cd/values.yaml
Comment (pdrastil) [reply charts/argo-cd/values.yaml:125]: I would like to keep this for refactoring and drop all `extraEnv` and `extraArgs` and keep just `args` and `env`.

File: charts/argo-cd/templates/argocd-server/deployment.yaml
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-server/deployment.yaml:37]: Done

File: charts/argo-cd/values.yaml
Comment (mkilchhofer) [reply charts/argo-cd/values.yaml:125]: Then we need to add something different. Currently the key is undocumented in README (line 404):
```
| controller.args | object | `{}` | |
```

Comment (pdrastil) [reply charts/argo-cd/values.yaml]: OK marked as deprecated and then we will change the meaning during the recatoring.

**Comment by pdrastil** [comment]
@mkilchhofer any update on this?

**Review by jmeridth — COMMENTED** [review]
LGTM 👍

nice cleanup of `if` to `with`

**Review by jmeridth — APPROVED** [review]

**Pull request merged**
